### DISCUSSION TITLE
How to pass a customized onChange handler to the onChange event for controlled fields

### DISCUSSION THREAD

**Discussion by arthu** [react-hook-form/react-hook-form#2194] [Q&A] [answered]
In v5.x, I could pass a onChange handler to the onChange event of the controlled field, which would trigger some customized onChange handler, and it worked perfectly.

...
const changeHandler = (value) => {
      const [v] = value;
      
        // call customized onChange logic for each field
        myChangeHandler(v, name);
      
      return v;
};
<Controller as={<FormField />} name={name} onFocus={focusHandler} onChange={changeHandler} {...props} />;

After I migrated over to version 6, this customized onChange handler was no longer being triggered
None of the following options worked
<Controller as={<FormField />} name={name} onFocus={focusHandler} onChange={changeHandler} {...props} />;
      or
<Controller
        name={name}
        onFocus={onFocus}
        onChange={changeHandler}
        rules={rules}
        render={(props) => <FormField onChange={props.onChange} {...newProps} />}
      />

I figured it might be the way I passed changeHandler was wrong, but don't know how to fix it. Can anyone shed some lights how I could pass a customized changeHandler to intercept the onChange event in v6 the same way as v5.x? Thanks a lot

**Comment by bluebill1049** [comment, marked as answer]
```
<Controller
  name={name}
  onFocus={onFocus}
  rules={rules}
  render={(props) => <FormField onChange={(value) => props.onChange(value); changeHandler(value); } {...newProps} />}
/>
```

You can do that inside the `render` prop.

**Reply by DuneArchitect** [reply]
@arthu could you help me understand what is `newProps` here?

**Reply by sergeyivanovv89** [reply]
That is pseudocode lol

**Reply by glc-mejd** [reply]
Does not work for me

**Reply by amanda-a-reis** [reply]
thank you! This helped me perfectly

**Reply by amit14apr** [reply]
I did tried with above solution..... but it is not working for me..  below is the code for reference.
import {
  useForm,
  Controller,
  FormProvider,
  useFormContext,
  } from "react-hook-form";
  
  const {
  control,
  formState: { errors },
  } = useFormContext();
  
  const [colorVal, setColor] = useState('#ffffff');
  
  return (
  <>
  Color Details
  <Controller
  control={control}
  name="ColorValue"
  rules={{
  required: "This field is required.",
  minLength: { value: 3, message: "Please enter minimum of 3 characters." }
  }}
  render={({ field }) => (
  <TextField
  id="ColorValue"
  variant="filled"
  label="Color Value *"
  size="small"
  onChange={(value) => {field.onChange(value); setColor(value);} } // onChange event is not working...
  InputProps={{
  endAdornment: (
  
    <InputAdornment position="end">
    <input type="color" value={colorVal} disabled/>
  </InputAdornment>
  
  ),
  }}
  {...field}
  error={Boolean(errors?.ColorValue)}
  helperText={errors.ColorValue?.message}
  />
  )}
  />

**Reply by Moshyfawn** [reply]
Are you sure MUI `TextField` returns the `value` as the first callback argument? I believe, it's the `event` object instead and the `value` you want to pass into your `field.onChange` handler is the second callback argument.

**Comment by arthu** [comment]
Works! Thank YOU Bill, really appreciate you help!!

```
<Controller
        name={name}
        onFocus={onFocus}
        rules={rules}
        render={(props) => (
          <FormField
            onChange={(value) => {
              props.onChange(value);
              changeHandler(value, name);
            }}
            value={props.value}
            {...newProps}
          />
        )}
      />
```

**Reply by makarenkovdo** [reply]
But how we can prevent rerender FormFiled (for example) ?
I need to wrap onChange on useCallback, but how?

**Comment by HadiSDev** [comment]
Hi I tried this: 


```
        <Controller
          name="TextField"
          control={control}
          
          render={(props) => <input type="number" value={props.value} onChange={(event) => {
            if(!isInt(event.target.value))
              {
                event.preventDefault()
              }
              else{
                props.onChange(event);
              }
            
          }}/>}
        />
```

But I got this error:

```
Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check the render method of `Controller`.
```

**Reply by HadiSDev** [reply]
Nvm it was an old version I used

**Reply by AlirezaM1997** [reply]
ok

**Comment by subrato-pattanaik** [comment]
@bluebill1049  
can you help with this question?
https://stackoverflow.com/q/64371532/13847690

**Reply by subrato-pattanaik** [reply]
If the user accidentally moves out of that page after inputting some data in the form, confirmation should be obtained from the user before moving out.
I want to achieve this.

**Reply by bluebill1049** [reply]
check `formState` with `isDirty` state and then you can pop up a confirm before user leavel.

**Comment by alanrezendeee** [comment]
I did it this way and it worked for me:

`<FormControl fullWidth={false} sx={{ width: '290px', mb: 6 }}>
  <Controller
    name='cnpj'
    control={control}
    render={(props) => (
      <TextField
        value={props.field.value}  
        label='Cnpj'
        onChange={(value): void => {
          props.field.onChange(value)
          changeHandler(value)
        }}
        placeholder='(e.g.: 60.133.365/0001-16)'
        error={Boolean(errors.cnpj)}
      />
    )}
  />
  {errors.cnpj && <FormHelperText sx={{ color: 'error.main' }}>{errors.cnpj.message}</FormHelperText>}
</FormControl>`


My method handleHandler:

`const changeHandler = (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
  setCnpjToSearch(event.target.value)
}`
